### PR TITLE
Home work 24

### PR DESCRIPTION
**Pull request by UnitOmega13**

### CODE REVIEW

File: src/src/main/java/controller/NewUserServlet.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        } else {
+            request.setAttribute("error", "Passwords are not equals");
+            request.setAttribute("email", email);
+            request.setAttribute("accessRole", accessRole);
+            request.setAttribute("action", "/users/add");
+            request.getRequestDispatcher("/registration.jsp").forward(request, response);
+        }
+    }
+
+    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
```
Comment (Sofasmile) [inline src/src/main/java/controller/NewUserServlet.java]: https://mate-academy.github.io/style-guides/java/java.html#s4.4-column-limit

File: src/src/main/java/controller/filter/UserAccessFilter.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import org.apache.log4j.Logger;
+
+import javax.servlet.*;
+import javax.servlet.annotation.WebFilter;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpSession;
+import java.io.IOException;
+
+@WebFilter(filterName = "UserAccessFilter",
+        value = {"/user/new", "/user/delete", "/user/edit"})
```
Comment (Sofasmile) [inline src/src/main/java/controller/filter/UserAccessFilter.java]: do this filter work correctly?

File: src/src/main/java/dao/ProductDAO.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package dao;
+
+import java.util.List;
+
+public interface ProductDAO<Product> {
+
+    void add(Product product);
+    List<Product> getAll();
+    model.Product getProduct(long productID);
```
Comment (Sofasmile) [inline src/src/main/java/dao/ProductDAO.java]:
```suggestion
    Product getProduct(long productID);
```

File: src/src/main/java/dao/ProductDAO.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package dao;
+
+import java.util.List;
+
+public interface ProductDAO<Product> {
```
Comment (Sofasmile) [inline src/src/main/java/dao/ProductDAO.java]: do we need generics here?

File: src/src/main/java/dao/UsersDAO.java [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+import java.util.List;
+
+public interface UsersDAO<User> {
+
+    void add(User user);
+    List<User> getAll();
+    User getUserById(long userId);
+    void removeUser(User user);
+    model.User getUserByEmail(String email);
```
Comment (Sofasmile) [inline src/src/main/java/dao/UsersDAO.java]:
```suggestion
    User getUserByEmail(String email);
```

File: src/src/main/java/service/UserService.java [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package service;
+
+import java.util.List;
+
+public interface UserService<User> {
+
+    void add(User user);
+    List<User> getAll();
+    User getUserById(long userID);
+    model.User getUserByEmail(String email);
```
Comment (Sofasmile) [inline src/src/main/java/service/UserService.java]:
```suggestion
    User getUserByEmail(String email);
```

Comment (UnitOmega13) [reply src/src/main/java/service/UserService.java]: fixed

File: src/src/main/java/controller/AccessServlet.java [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+@WebServlet(name = "AccessServlet", value = "/access")
+public class AccessServlet extends HttpServlet {
+
+    protected void doGet(HttpServletRequest request,
+                         HttpServletResponse response) throws ServletException, IOException {
+        request.setAttribute("accessDenied", "Not enough rights");
+        request.getRequestDispatcher("/users").forward(request, response);
```
Comment (Sofasmile) [inline src/src/main/java/controller/AccessServlet.java]: if user hasn't enough rights, I think that we shouldn't redirect it to /users page

File: src/src/main/java/controller/ChangeProductServlet.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.util.logging.Logger;
+
+@WebServlet(value = "/products/edit")
+public class ChangeProductServlet extends HttpServlet {
+
+    private static final Logger LOGGER = Logger.getLogger(String.valueOf(ChangeProductServlet.class));
+    private static final ProductService productService = ProductServiceFactory.getInstance();
+    Product product;
```
Comment (Sofasmile) [inline src/src/main/java/controller/ChangeProductServlet.java:23]: try doesn't use this variable

File: src/src/main/java/controller/ChangeUserServlet.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        String email = request.getParameter("email");
+        String accessRole = request.getParameter("accessRole");
+        String password = request.getParameter("password");
+        String reEnteredPassword = request.getParameter("repeatPassword");
+        Optional<User> optionalUser = userService.getUserById(userId);
+        if (optionalUser.isPresent() &&
+                password.equals(reEnteredPassword) &&
+                !password.equals("")) {
+            User user = optionalUser.get();
+            LOGGER.info("user " + user + " edited");
```
Comment (Sofasmile) [inline src/src/main/java/controller/ChangeUserServlet.java]: user will be edited after setting all changes

File: src/src/main/java/controller/filter/UserAccessFilter.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
+package controller.filter;
+
+import controller.ChangeProductServlet;
+import model.User;
+import org.apache.log4j.Logger;
+
+import javax.servlet.*;
```
Comment (Sofasmile) [inline src/src/main/java/controller/filter/UserAccessFilter.java:7]: :)

File: src/src/main/java/dao/ProductDAO.java (hunk 3 above)
Comment (Sofasmile) [reply src/src/main/java/dao/ProductDAO.java]: ? what about previous comment?

**Review by Sofasmile — COMMENTED** [review]
add id here for users and products

**Pull request closed without merging**